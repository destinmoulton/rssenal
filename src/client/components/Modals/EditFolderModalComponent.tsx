import debug from "debug";
import * as React from "react";

import { Button, Header, Modal } from "semantic-ui-react";

import { IFolder } from "../../types";

const log = debug("rssenal:EditFolderModalComponent");

export interface IEditFolderModalMapDispatch {
    folderSave: (folder: any) => void;
}

interface IEditFolderModalProps {
    isModalOpen: boolean;
    onCloseModal: () => void;
    folder: IFolder;
}

type TAllProps = IEditFolderModalMapDispatch & IEditFolderModalProps;

interface IState {
    folderInfo: any;
}

const INITIAL_FOLDER = { name: "", _id: "" };

class EditFolderModalComponent extends React.Component<TAllProps, IState> {
    state: IState = {
        folderInfo: INITIAL_FOLDER
    };

    static getDerivedStateFromProps(props: TAllProps, state: IState) {
        if (props.folder) {
            return {
                folderInfo: props.folder
            };
        }

        return null;
    }

    _handleClose = () => {
        log("_handleClose()");
        this.setState({
            folderInfo: INITIAL_FOLDER
        });
        this.props.onCloseModal();
    };

    _handleSave = () => {
        log("_handleSave()");
        const { folderInfo } = this.state;

        if (folderInfo.name.trim() !== "") {
            this.props.folderSave(folderInfo);
            this._handleClose();
        }
    };

    _handleInputKeyPress = (e: any) => {
        if (e.key === "Enter") {
            this._handleSave();
        }
    };

    _handleInputOnChange = (e: any) => {
        const { folderInfo } = this.state;
        folderInfo.name = e.target.value;
        this.setState({
            folderInfo
        });
    };

    _buildEditInput() {
        const { folderInfo } = this.state;

        return (
            <input
                autoFocus
                value={folderInfo.name}
                placeholder="Folder name..."
                onKeyPress={this._handleInputKeyPress}
                onChange={this._handleInputOnChange}
                className="rss-foldermodal-name"
            />
        );
    }

    _buildButtons(saveText: string) {
        return (
            <div>
                <Button
                    color="orange"
                    content="Cancel"
                    floated="left"
                    icon="cancel"
                    inverted
                    onClick={this._handleClose}
                />
                <Button
                    color="green"
                    content={saveText}
                    icon="checkmark"
                    inverted
                    onClick={this._handleSave}
                />
            </div>
        );
    }

    render() {
        log("render()");
        const { folder, isModalOpen } = this.props;
        const form = this._buildEditInput();

        let modalHeader = "Add Folder";

        let saveText = "Create";
        if (folder && folder._id !== "") {
            modalHeader = "Edit Folder";
            saveText = "Save";
        }

        const buttons = this._buildButtons(saveText);
        return (
            <Modal
                open={isModalOpen}
                onClose={this._handleClose}
                size="tiny"
                className="rss-modal"
            >
                <Header icon="plus" content={modalHeader} />
                <Modal.Content>{form}</Modal.Content>
                <Modal.Actions>{buttons}</Modal.Actions>
            </Modal>
        );
    }
}

export default EditFolderModalComponent;
