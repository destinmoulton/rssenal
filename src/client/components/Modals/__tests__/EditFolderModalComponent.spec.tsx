import * as React from "react";

import { shallow } from "enzyme";

import EditFolderModalComponent from "../EditFolderModalComponent";

import folder from "../../../../../test/data/folder";

describe("<EditFolderModalComponent />", () => {
    describe("renders and matches the snapshots", () => {
        it("matches the snapshot for Add New Folder", () => {
            const wrapper = shallow(
                <EditFolderModalComponent
                    folder={{ _id: "", name: "", order: 0 }}
                    isModalOpen={true}
                    onCloseModal={jest.fn()}
                    beginSaveFolder={jest.fn()}
                />
            );
            expect(wrapper).toMatchSnapshot();
        });

        it("matches the snapshot for Edit Existing Folder", () => {
            const wrapper = shallow(
                <EditFolderModalComponent
                    folder={folder}
                    isModalOpen={true}
                    onCloseModal={jest.fn()}
                    beginSaveFolder={jest.fn()}
                />
            );
            expect(wrapper).toMatchSnapshot();
        });
    });
});
