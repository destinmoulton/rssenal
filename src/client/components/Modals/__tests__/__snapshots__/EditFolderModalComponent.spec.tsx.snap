// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditFolderModalComponent /> renders and matches the snapshots matches the snapshot for Add New Folder 1`] = `
<Modal
  centered={true}
  className="rss-modal"
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
  onClose={[Function]}
  open={true}
  size="tiny"
>
  <Header
    content="Add Folder"
    icon="plus"
  />
  <ModalContent>
    <input
      autoFocus={true}
      className="rss-foldermodal-name"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Folder name..."
      value=""
    />
  </ModalContent>
  <ModalActions>
    <div>
      <Button
        as="button"
        color="orange"
        content="Cancel"
        floated="left"
        icon="cancel"
        inverted={true}
        onClick={[Function]}
        role="button"
      />
      <Button
        as="button"
        color="green"
        content="Create"
        icon="checkmark"
        inverted={true}
        onClick={[Function]}
        role="button"
      />
    </div>
  </ModalActions>
</Modal>
`;

exports[`<EditFolderModalComponent /> renders and matches the snapshots matches the snapshot for Edit Existing Folder 1`] = `
<Modal
  centered={true}
  className="rss-modal"
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
  onClose={[Function]}
  open={true}
  size="tiny"
>
  <Header
    content="Edit Folder"
    icon="plus"
  />
  <ModalContent>
    <input
      autoFocus={true}
      className="rss-foldermodal-name"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Folder name..."
      value="Random"
    />
  </ModalContent>
  <ModalActions>
    <div>
      <Button
        as="button"
        color="orange"
        content="Cancel"
        floated="left"
        icon="cancel"
        inverted={true}
        onClick={[Function]}
        role="button"
      />
      <Button
        as="button"
        color="green"
        content="Save"
        icon="checkmark"
        inverted={true}
        onClick={[Function]}
        role="button"
      />
    </div>
  </ModalActions>
</Modal>
`;
